### ISSUE TITLE
Allow MultiplayerSynchronizer hidding what client has authority, pretending the server does.

### ISSUE THREAD

**Issue by Pshy0** [godotengine/godot-proposals#8646]
### Describe the project you are working on

A multiplayer game with an authoritative server that uses `MultiplayerSynchronizer`.

### Describe the problem or limitation you are having in your project

I want to improve my game's security by preventing clients from knowing other clients are connected, also preventing them from communicating with each other, or from identifying which pawn is a player and which is a dummy or bot.

An obvious solution would be to disable `multiplayer.server_relay`.
However, this prevents `MultiplayerSynchronizer`'s from working, as the server will see properties being edited by clients, but clients won't see properties edited by each other.

Also, even if I only wanted to prevent clients from making rpc calls to each other, configuring rpc functions with `@rpc("authority")` would not be enough, as two modded clients could have misconfigured rpc functions.
This could for instance allow two modded clients exploiting this to turn my server into a relay for arbitrary data I do not approve relaying.
So in this case, `multiplayer.server_relay` would improve security, but is incompatible with using `MultiplayerSynchronizer`.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

An option could be added to synchronizers (or to `multiplayer`) to make the server-side `MultiplayerSynchronizer` replicate data from a client that have authority on that data to other clients, like if for other clients, the server was the authority.

This way, the server knows what client has authority on what node, but does not reveal it to clients.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Imagine a player pawn `Player2` controlled by client with id 2, and connected to a server with many other clients.
Each PlayerN node has a MultiplayerSynchronizer.

Without this feature, `multiplayer.server_relay` needs to be enabled for this setup to work.
Player2's MultiplayerSynchronizer authority would be set to 2 on client 2, the server, and all other clients that need properties updated.
As a consequence, clients would know about each other, and they will even have knowledge of what pawn is controlled by a player and what pawn is not (cf a bot).

With this feature, `multiplayer.server_relay` could be disabled, improving overall security.
Player2's MultiplayerSynchronizer authority would be set to 2 on client 2 and the server, but to 1 on other clients.
The server would accept updates from the authority client 2, and then relay them to other clients as being updated by 1, like if the server was the real authority.
This way, clients do not know about each other, while `MultiplayerSynchronizer` still function.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

A way would be to make the server fully authoritative, and expecting player inputs instead of players updating their positions/speed. But this would cause further latency, this is not acceptable for all games.

The following allows disabling `multiplayer.server_relay` and still use synchronizers:

The pawns node could have two synchronizers configured with the same properties.
On the server, one's authority would be the client, the other would be the server.
On the client controlling the pawn, one's authority would be itself, and the other would be disabled.
On other clients, one would be disabled, the other's authority would be set to the server.

Disabling visibility for the client that receive the error on the server-to-client synchronizer have no effect, and the synchronizer keep synchronizing properties (is this a godot bug?). So to disable synchronizers, I set their `delta_time` to an high value to prevent them from sending updates, and I set their authority to a non-existing id to prevent them from considering received updates valid.

This is the code to change who controls a pawn (called on both client and server):
```gdscript
func set_controller_peer(peer_id: int):
	if peer_id != 1 and multiplayer.get_unique_id() != 1 and peer_id != multiplayer.get_unique_id():
		debug.push_warning("set_controller_peer called on client with the id of another client")
		peer_id = 1
	if peer_id == 1:
		$server_to_client_synchronizer.set_multiplayer_authority(1, false)
		$server_to_client_synchronizer.replication_interval = 0.0
		$server_to_client_synchronizer.delta_interval = 0.0
		$client_to_server_synchronizer.set_multiplayer_authority(-1, false)
		$client_to_server_synchronizer.replication_interval = 999999
		$client_to_server_synchronizer.delta_interval = 999999
		if multiplayer.get_unique_id() == 1:
			$server_to_client_synchronizer.set_visibility_for($server_to_client_synchronizer.get_multiplayer_authority(), true)
			$server_to_client_synchronizer.set_visibility_for(peer_id, true)
	elif peer_id != 1 and multiplayer.get_unique_id() == 1:
		$server_to_client_synchronizer.set_multiplayer_authority(1, false)
		$server_to_client_synchronizer.replication_interval = 0.0
		$server_to_client_synchronizer.delta_interval = 0.0
		$client_to_server_synchronizer.set_multiplayer_authority(peer_id, false)
		$client_to_server_synchronizer.replication_interval = 0.0
		$client_to_server_synchronizer.delta_interval = 0.0
		$server_to_client_synchronizer.set_visibility_for($server_to_client_synchronizer.get_multiplayer_authority(), true)
		$server_to_client_synchronizer.set_visibility_for(peer_id, false)
		# from now on, `$server_to_client_synchronizer.get_visibility_for(peer_id)` returns false, but server_to_client_synchronizer still attempts to synchronize properties :v
	elif peer_id == multiplayer.get_unique_id():
		$server_to_client_synchronizer.set_multiplayer_authority(-1, false)
		$server_to_client_synchronizer.replication_interval = 999999
		$server_to_client_synchronizer.delta_interval = 999999
		$client_to_server_synchronizer.set_multiplayer_authority(peer_id, false)
		$client_to_server_synchronizer.replication_interval = 0.0
		$client_to_server_synchronizer.delta_interval = 0.0
	else:
		assert(false, "unreachable")
```

While this works, this also causes a huge spam of the following error message, due to `server_to_client_synchronizer` synchronizing to clients for which visibility is disabled:
> E 0:00:10:0794   on_delta_receive: Ignoring delta for non-authority or invalid synchronizer.

### Is there a reason why this should be core and not an add-on in the asset library?

This would encourage making more secure games, thanks to only having to change one setting once, rather than having to handle a complex dual-synchronizer system or anything else.

**Comment by copygirl** [comment]
I want to comment that I have essentially the same issue, but the title is a bit wordy. In essence, what this issue is asking for is for `MultiplayerSynchronizer` to somehow work with `server_relay` disabled, right? Perhaps the title could be changed to reflect this?

Some more of my thoughts:

On one hand, I could guess that `MultiplayerSynchronizer` works through relaying the synced properties in a sense from one client to another, so it should respect the `server_relay` setting.

On the other hand, it is such a useful, straightforward, simple and less error-prone way to synchronize properties between objects, including ones controlled by players, so I feel like it should continue working even when `server_relay` is disabled. After all, you have set up explicitly which properties are valid to be synced. For security purposes, there could be a signal that validates the properties and drops anything suspicious before getting applied or synced to other players. This is useful either way. (But should be addressed by another issue.)
